### PR TITLE
chore: refactor cross domain communications

### LINKED ISSUE #16777: Multi-domain: Create postMessage bus for cross-domain communication
There's currently rudimentary postMessage communication implemented, but it needs to be made more robust with a proper API and serialization handling.

- Move internal multidomain-related events off the public Cypress/cy event bus onto a dedicated internal bus
- Create a naming scheme for events sent over the bus
- ~Create a promise-based API for request-response~
- ~Ensure serialization of more complex of data works, or else implemented JSON stringify and parsing~

### PR DESCRIPTION
**Pull request by AtofStryker**
- Closes #16777 

### Additional details

The effort here is to consolidate some of the multi domain logic to be more contained and isolated. 

The multi domain events on the primary domain have been moved off the public `Cypress.action` event bus into a message handler that exists on the `PrimaryDomainCommunicator`. 

A similar approach has been used on the spec bridge side, where events have been moved out of the spec bridge (`multidomain/index.js`) over to the `SpecBridgeDomainCommunicator`.

When communicating with each other, the event prefix `cross:domain:` is used to filter out any other extraneous noise that might also be used by other domains interacting with `postMessage`

With both handlers now living in their respective classes, `postMessage` implementation details are now abstracted and go through a common interface. The primary domain now communicates to the spec bridge via the `toSpecBridge` method, and contrarily the spec bridge communicates back to the primary domain through the `toPrimary` method. 

### PR Tasks

- [n/a] Have tests been added/updated?
- [n/a ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [n/a] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/runner-shared/src/event-manager.js [hunk 1]
```diff
@@ -26,9 +26,9 @@ const driverToReporterEvents = 'paused session:add'.split(' ')
 const driverToLocalAndReporterEvents = 'run:start run:end'.split(' ')
 const driverToSocketEvents = 'backend:request automation:request mocha recorder:frame'.split(' ')
 const driverTestEvents = 'test:before:run:async test:after:run'.split(' ')
-const driverToLocalEvents = 'viewport:changed config stop url:changed page:loading visit:failed visit:blank expect:domain'.split(' ')
```
Comment (AtofStryker) [inline packages/runner-shared/src/event-manager.js:29]: `expect:domain` now lives on the primary communicator's event bus

File: packages/runner-shared/src/event-manager.js [hunk 2]
```diff
@@ -26,9 +26,9 @@ const driverToReporterEvents = 'paused session:add'.split(' ')
...
+const driverToLocalEvents = 'viewport:changed config stop url:changed page:loading visit:failed visit:blank'.split(' ')
 const socketRerunEvents = 'runner:restart watched:file:changed'.split(' ')
-const socketToDriverEvents = 'net:stubbing:event request:event script:error cross:domain:html:received'.split(' ')
```
Comment (AtofStryker) [inline packages/runner-shared/src/event-manager.js:31]: the primary communicator now listens to this event on the web socket directly to offload the work from the public event bus

File: packages/driver/src/multidomain/communicator.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        } else {
+          this.emit(messageName, data.data)
+        }
+
+        return
+      }
+
+      // TODO: how do we want to log unexpected messages, if at all?
+      // eslint-disable-next-line no-console
+      console.log('Unexpected postMessage:', data)
```
Comment (lmiller1990) [inline packages/driver/src/multidomain/communicator.ts:53]: Perhaps like this using `debug` https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cy/commands/navigation.ts#L14-L15

File: cli/types/cypress.d.ts [hunk 4]
```diff
@@ -5479,6 +5490,13 @@ declare namespace Cypress {
      * @see https://on.cypress.io/catalog-of-events#App-Events
      */
     (action: 'window:load', fn: (win: AUTWindow) => void): Cypress
+    /**
+     * Fires when 'window:load' fires, as well as when a page on a secondary
+     * domain loads.
+     *
+     * WARNING: Intended for internal use. Subject to change or be removed.
+     */
+    (action: 'internal:window:load', fn: (details: InternalWindowLoadDetails) => void): Cypress
```
Comment (tbiethman) [inline cli/types/cypress.d.ts:5499]: Can we extend the interface internally where we need to with interface merging and not expose this here?

File: packages/driver/src/cypress/log.ts [hunk 5]
```diff
@@ -181,8 +177,10 @@ const defaults = function (state, config, obj) {
     return t._currentRetry || 0
   }
 
+  counter++
+
   _.defaults(obj, {
-    id: (counter += 1),
+    id: `log-${window.location.origin}-${counter}`,
```
Comment (tbiethman) [inline packages/driver/src/cypress/log.ts:183]: Does this correct the issue reported in https://github.com/cypress-io/cypress/issues/19545?

**Review by tbiethman — APPROVED** [review]

**Comment by tbiethman** [comment]
This refactor looks good :+1: I'll spend some time reviewing the multidomain feature branch so I can give deeper feedback on subsequent PRs in this space.

File: packages/driver/src/cypress/log.ts (hunk 5 above)
Comment (chrisbreiding) [reply packages/driver/src/cypress/log.ts:183]: No, unfortunately not.

File: cli/types/cypress.d.ts (hunk 4 above)
Comment (chrisbreiding) [reply cli/types/cypress.d.ts]: Yes indeed, good idea. Addressed in [`157ea09`](https://github.com/cypress-io/cypress/pull/19430/commits/157ea0907a61600f33fef7a0ced09f80e0fd479f).

**Review by ryanthemanuel — APPROVED** [review]

**Pull request merged**